File: hunter_fitness_api/Models/Hunter.cs
using System.ComponentModel.DataAnnotations;

namespace HunterFitness.API.Models
{
    public class Hunter
    {
        [Key]
        public Guid HunterID { get; set; } = Guid.NewGuid();

        [Required]
        [StringLength(50)]
        public string Username { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        [EmailAddress]
        public string Email { get; set; } = string.Empty;

        [Required]
        [StringLength(255)]
        public string PasswordHash { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        public string HunterName { get; set; } = string.Empty;

        // Stats del Hunter (Solo Leveling style)
        public int Level { get; set; } = 1;
        public int CurrentXP { get; set; } = 0;
        public int TotalXP { get; set; } = 0;

        [StringLength(10)]
        public string HunterRank { get; set; } = "E";

        // Stats principales
        public int Strength { get; set; } = 10;
        public int Agility { get; set; } = 10;
        public int Vitality { get; set; } = 10;
        public int Endurance { get; set; } = 10;

        // Progreso y streaks
        public int DailyStreak { get; set; } = 0;
        public int LongestStreak { get; set; } = 0;
        public int TotalWorkouts { get; set; } = 0;

        // Metadatos
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? LastLoginAt { get; set; }
        public bool IsActive { get; set; } = true;

        [StringLength(500)]
        public string? ProfilePictureUrl { get; set; }

        // Navigation Properties - CORREGIDO: Nombres correctos
        public virtual ICollection<HunterDailyQuest> DailyQuests { get; set; } = new List<HunterDailyQuest>();
        public virtual ICollection<DungeonRaid> DungeonRaids { get; set; } = new List<DungeonRaid>();
        public virtual ICollection<HunterAchievement> Achievements { get; set; } = new List<HunterAchievement>();
        public virtual ICollection<HunterEquipment> Equipment { get; set; } = new List<HunterEquipment>();
        public virtual ICollection<QuestHistory> QuestHistory { get; set; } = new List<QuestHistory>();

        // Métodos Helper
        public int GetTotalStatsWithEquipment()
        {
            var baseStats = Strength + Agility + Vitality + Endurance;
            
            if (Equipment?.Any() != true)
                return baseStats;

            var equipmentBonus = Equipment
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.StrengthBonus + e.Equipment.AgilityBonus + 
                         e.Equipment.VitalityBonus + e.Equipment.EnduranceBonus);
            
            return baseStats + equipmentBonus;
        }

        public string GetRankDisplayName()
        {
            return HunterRank switch
            {
                "E" => "Rookie Hunter",
                "D" => "Bronze Hunter", 
                "C" => "Silver Hunter",
                "B" => "Gold Hunter",
                "A" => "Elite Hunter",
                "S" => "Master Hunter",
                "SS" => "Legendary Hunter",
                "SSS" => "Shadow Monarch",
                _ => "Unknown Rank"
            };
        }

        public int GetXPRequiredForNextLevel()
        {
            // Curva exponencial para leveling
            return (int)(100 * Math.Pow(1.5, Level - 1));
        }

        public bool CanLevelUp()
        {
            return CurrentXP >= GetXPRequiredForNextLevel();
        }

        public string GetNextRankRequirement()
        {
            return HunterRank switch
            {
                "E" => "Reach Level 11 to become Bronze Hunter",
                "D" => "Reach Level 21 to become Silver Hunter",
                "C" => "Reach Level 36 to become Gold Hunter", 
                "B" => "Reach Level 51 to become Elite Hunter",
                "A" => "Reach Level 71 to become Master Hunter",
                "S" => "Reach Level 86 to become Legendary Hunter",
                "SS" => "Reach Level 96 to become Shadow Monarch",
                "SSS" => "Maximum rank achieved!",
                _ => "Unknown rank progression"
            };
        }

        public void UpdateRankBasedOnLevel() // Asegúrate que esto se llame si el nivel cambia
        {
            string previousRank = HunterRank;
            HunterRank = Level switch
            {
                >= 96 => "SSS",
                >= 86 => "SS", 
                >= 71 => "S",
                >= 51 => "A",
                >= 36 => "B",
                >= 21 => "C",
                >= 11 => "D",
                _ => "E"
            };
            // Opcional: Log si el rango cambió
            if (HunterRank != previousRank) {
                Console.WriteLine($"Hunter {HunterName} ranked up to {HunterRank}!");
            }
        }

        public void ProcessLevelUp()
        {
            bool leveledUpThisCycle = false;
            while (CanLevelUp())
            {
                var xpRequired = GetXPRequiredForNextLevel();
                CurrentXP -= xpRequired;
                Level++;
                leveledUpThisCycle = true; // Marcamos que al menos un nivel se subió
            }
            // Actualizar el rango solo si realmente hubo un cambio de nivel en este ciclo.
            if (leveledUpThisCycle)
            {
                UpdateRankBasedOnLevel();
            }
        }

        public decimal GetLevelProgressPercentage()
        {
            var xpRequired = GetXPRequiredForNextLevel();
            return xpRequired > 0 ? (decimal)CurrentXP / xpRequired * 100 : 0;
        }

        // Stats totales con equipment
        public int GetTotalStrength()
        {
            return Strength + GetEquipmentStrengthBonus();
        }

        public int GetTotalAgility()
        {
            return Agility + GetEquipmentAgilityBonus();
        }

        public int GetTotalVitality()
        {
            return Vitality + GetEquipmentVitalityBonus();
        }

        public int GetTotalEndurance()
        {
            return Endurance + GetEquipmentEnduranceBonus();
        }

        // Métodos para bonificaciones de equipment
        public int GetEquipmentStrengthBonus()
        {
            return Equipment?
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.StrengthBonus) ?? 0;
        }

        public int GetEquipmentAgilityBonus()
        {
            return Equipment?
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.AgilityBonus) ?? 0;
        }

        public int GetEquipmentVitalityBonus()
        {
            return Equipment?
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.VitalityBonus) ?? 0;
        }

        public int GetEquipmentEnduranceBonus()
        {
            return Equipment?
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.EnduranceBonus) ?? 0;
        }

        public decimal GetTotalXPMultiplier()
        {
            if (Equipment?.Any() != true)
                return 1.0m;

            var multiplier = Equipment
                .Where(e => e.IsEquipped && e.Equipment != null)
                .Sum(e => e.Equipment!.XPMultiplier - 1.0m);

            return 1.0m + multiplier;
        }

        // Validaciones
        public bool IsEligibleForRankUp()
        {
            return HunterRank switch
            {
                "E" => Level >= 11,
                "D" => Level >= 21,
                "C" => Level >= 36,
                "B" => Level >= 51,
                "A" => Level >= 71,
                "S" => Level >= 86,
                "SS" => Level >= 96,
                "SSS" => false, // Máximo rank
                _ => false
            };
        }

        public int GetDaysSinceJoining()
        {
            return Math.Max(1, (DateTime.UtcNow - CreatedAt).Days);
        }

        public decimal GetAverageXPPerDay()
        {
            var days = GetDaysSinceJoining();
            return days > 0 ? (decimal)TotalXP / days : 0;
        }

        public bool IsNewHunter()
        {
            return GetDaysSinceJoining() <= 7;
        }

        public bool IsVeteranHunter()
        {
            return GetDaysSinceJoining() >= 365;
        }

        // Override para mejor debugging
        public override string ToString()
        {
            return $"Hunter: {HunterName} (Level {Level} {HunterRank}) - {Username}";
        }

        // Método para validar integridad de datos
        public List<string> ValidateData()
        {
            var errors = new List<string>();

            if (string.IsNullOrWhiteSpace(Username))
                errors.Add("Username is required");

            if (string.IsNullOrWhiteSpace(Email))
                errors.Add("Email is required");

            if (string.IsNullOrWhiteSpace(HunterName))
                errors.Add("Hunter name is required");

            if (Level < 1)
                errors.Add("Level must be at least 1");

            if (CurrentXP < 0)
                errors.Add("Current XP cannot be negative");

            if (TotalXP < CurrentXP)
                errors.Add("Total XP cannot be less than current XP");

            if (Strength < 1 || Agility < 1 || Vitality < 1 || Endurance < 1)
                errors.Add("All stats must be at least 1");

            if (DailyStreak > LongestStreak)
                errors.Add("Daily streak cannot be longer than longest streak");

            var validRanks = new[] { "E", "D", "C", "B", "A", "S", "SS", "SSS" };
            if (!validRanks.Contains(HunterRank))
                errors.Add("Invalid hunter rank");

            return errors;
        }
    }
}

File: hunter_fitness_app/lib/models/hunter.dart
class Hunter {
  final String hunterID;
  final String username;
  final String email;
  final String hunterName;
  final int level;
  final int currentXP;
  final int totalXP;
  final String hunterRank;
  final String rankDisplayName;
  final int strength;
  final int agility;
  final int vitality;
  final int endurance;
  final int totalStats;
  final int dailyStreak;
  final int longestStreak;
  final int totalWorkouts;
  final int xpRequiredForNextLevel;
  final double levelProgressPercentage;
  final bool canLevelUp;
  final String nextRankRequirement;
  final DateTime createdAt;
  final DateTime? lastLoginAt;
  final String? profilePictureUrl;
  final List<EquippedItem> equippedItems;
  final Map<String, dynamic> additionalStats;

  Hunter({
    required this.hunterID,
    required this.username,
    required this.email,
    required this.hunterName,
    required this.level,
    required this.currentXP,
    required this.totalXP,
    required this.hunterRank,
    required this.rankDisplayName,
    required this.strength,
    required this.agility,
    required this.vitality,
    required this.endurance,
    required this.totalStats,
    required this.dailyStreak,
    required this.longestStreak,
    required this.totalWorkouts,
    required this.xpRequiredForNextLevel,
    required this.levelProgressPercentage,
    required this.canLevelUp,
    required this.nextRankRequirement,
    required this.createdAt,
    this.lastLoginAt,
    this.profilePictureUrl,
    required this.equippedItems,
    required this.additionalStats,
  });

  factory Hunter.fromJson(Map<String, dynamic> json) {
    return Hunter(
      hunterID: json['hunterID'] ?? '',
      username: json['username'] ?? '',
      email: json['email'] ?? '',
      hunterName: json['hunterName'] ?? '',
      level: json['level'] ?? 1,
      currentXP: json['currentXP'] ?? 0,
      totalXP: json['totalXP'] ?? 0,
      hunterRank: json['hunterRank'] ?? 'E',
      rankDisplayName: json['rankDisplayName'] ?? 'Rookie Hunter',
      strength: json['strength'] ?? 10,
      agility: json['agility'] ?? 10,
      vitality: json['vitality'] ?? 10,
      endurance: json['endurance'] ?? 10,
      totalStats: json['totalStats'] ?? 40,
      dailyStreak: json['dailyStreak'] ?? 0,
      longestStreak: json['longestStreak'] ?? 0,
      totalWorkouts: json['totalWorkouts'] ?? 0,
      xpRequiredForNextLevel: json['xpRequiredForNextLevel'] ?? 100,
      levelProgressPercentage: (json['levelProgressPercentage'] ?? 0.0).toDouble(),
      canLevelUp: json['canLevelUp'] ?? false,
      nextRankRequirement: json['nextRankRequirement'] ?? '',
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      lastLoginAt: json['lastLoginAt'] != null ? DateTime.parse(json['lastLoginAt']) : null,
      profilePictureUrl: json['profilePictureUrl'],
      equippedItems: (json['equippedItems'] as List<dynamic>? ?? [])
          .map((item) => EquippedItem.fromJson(item))
          .toList(),
      additionalStats: json['additionalStats'] ?? {},
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'hunterID': hunterID,
      'username': username,
      'email': email,
      'hunterName': hunterName,
      'level': level,
      'currentXP': currentXP,
      'totalXP': totalXP,
      'hunterRank': hunterRank,
      'rankDisplayName': rankDisplayName,
      'strength': strength,
      'agility': agility,
      'vitality': vitality,
      'endurance': endurance,
      'totalStats': totalStats,
      'dailyStreak': dailyStreak,
      'longestStreak': longestStreak,
      'totalWorkouts': totalWorkouts,
      'xpRequiredForNextLevel': xpRequiredForNextLevel,
      'levelProgressPercentage': levelProgressPercentage,
      'canLevelUp': canLevelUp,
      'nextRankRequirement': nextRankRequirement,
      'createdAt': createdAt.toIso8601String(),
      'lastLoginAt': lastLoginAt?.toIso8601String(),
      'profilePictureUrl': profilePictureUrl,
      'equippedItems': equippedItems.map((item) => item.toJson()).toList(),
      'additionalStats': additionalStats,
    };
  }

  String getRankIcon() {
    switch (hunterRank) {
      case 'E':
        return '🔰';
      case 'D':
        return '🥉';
      case 'C':
        return '🥈';
      case 'B':
        return '🥇';
      case 'A':
        return '💎';
      case 'S':
        return '👑';
      case 'SS':
        return '⭐';
      case 'SSS':
        return '🏹';
      default:
        return '🔰';
    }
  }

  String getMotivationalMessage() {
    if (canLevelUp) {
      return "Ready to level up! 🎉";
    } else if (levelProgressPercentage > 75) {
      return "Almost there, Hunter! 🔥";
    } else if (dailyStreak > 5) {
      return "Incredible streak! 💪";
    } else {
      return "Keep pushing forward! ⚔️";
    }
  }
}

class EquippedItem {
  final String equipmentID;
  final String itemName;
  final String itemType;
  final String rarity;
  final String rarityColor;
  final int strengthBonus;
  final int agilityBonus;
  final int vitalityBonus;
  final int enduranceBonus;
  final double xpMultiplier;
  final String statBonusDescription;
  final String? iconUrl;
  final int powerLevel;

  EquippedItem({
    required this.equipmentID,
    required this.itemName,
    required this.itemType,
    required this.rarity,
    required this.rarityColor,
    required this.strengthBonus,
    required this.agilityBonus,
    required this.vitalityBonus,
    required this.enduranceBonus,
    required this.xpMultiplier,
    required this.statBonusDescription,
    this.iconUrl,
    required this.powerLevel,
  });

  factory EquippedItem.fromJson(Map<String, dynamic> json) {
    return EquippedItem(
      equipmentID: json['equipmentID'] ?? '',
      itemName: json['itemName'] ?? '',
      itemType: json['itemType'] ?? '',
      rarity: json['rarity'] ?? 'Common',
      rarityColor: json['rarityColor'] ?? '#9E9E9E',
      strengthBonus: json['strengthBonus'] ?? 0,
      agilityBonus: json['agilityBonus'] ?? 0,
      vitalityBonus: json['vitalityBonus'] ?? 0,
      enduranceBonus: json['enduranceBonus'] ?? 0,
      xpMultiplier: (json['xpMultiplier'] ?? 1.0).toDouble(),
      statBonusDescription: json['statBonusDescription'] ?? '',
      iconUrl: json['iconUrl'],
      powerLevel: json['powerLevel'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'equipmentID': equipmentID,
      'itemName': itemName,
      'itemType': itemType,
      'rarity': rarity,
      'rarityColor': rarityColor,
      'strengthBonus': strengthBonus,
      'agilityBonus': agilityBonus,
      'vitalityBonus': vitalityBonus,
      'enduranceBonus': enduranceBonus,
      'xpMultiplier': xpMultiplier,
      'statBonusDescription': statBonusDescription,
      'iconUrl': iconUrl,
      'powerLevel': powerLevel,
    };
  }

  String getTypeIcon() {
    switch (itemType) {
      case 'Weapon':
        return '⚔️';
      case 'Armor':
        return '🛡️';
      case 'Accessory':
        return '💍';
      default:
        return '⚡';
    }
  }
}